### PR TITLE
Work around warning from `late_bound_lifetime_arguments` lint

### PR DESCRIPTION
**Pull request by lqd**
The generic parsing functions require lifetime annotations to build because of https://github.com/rust-lang/rust/issues/42508. 

However, this creates a warning from the `late_bound_lifetime_arguments` compatibility lint, described in https://github.com/rust-lang/rust/issues/42868, because they have both early-bound and late-bound lifetime parameters.

This PR works around this warning, in a functional but sub-optimal way. The result is quite verbose, and I didn't manage to "convince" rustc to think all lifetimes were early-bound at the `fn` definitions, but manages to avoid triggering both https://github.com/rust-lang/rust/issues/42508 and the lint, by passing these closures via variables at each call-site.

Fixes https://github.com/servo/rust-cssparser/issues/254

### CODE REVIEW

**Comment by emilio** [comment]
Is there any good reason to do this other than building warning-free?

I don't think we should adapt your code for this necessarily, unless there's a very good reason to do this. Doubly so if this makes the code uglier tbf...

**Comment by lqd** [comment]
Yes building without the warning was the reason for the change, as it's not clear when this is going to become an error.

**Comment by emilio** [comment]
As long as this is working around a rust bug I don't think they can reasonably turn it into an error...

**Comment by lqd** [comment]
Ah but this PR is not working around a rustc bug, it is "fixing" rust-cssparser's workaround for the rustc bug and which generates a compatiblity lint warning.

(the code is not that ugly on current stable. I haven't tested older stables so I don't know which one fixed the scope requirement, I only added it because CI fails on 1.31. These, and the existing ones, eg in parser.rs IIRC, could be removed on newer versions)

**Comment by SimonSapin** [comment]
The code comment is useful, but perhaps the whole explanation does not need to be repeated every time. Maybe move it to a github comment in https://github.com/servo/rust-cssparser/issues/254, and have the code link there?

With that change I feel the ugliness would not be *too* bad. I think warning-freedom is valuable, although in this case we aren’t actively doing a lot of work on this crate so warnings haven’t been much of a bother.

File: src/rules_and_declarations.rs [hunk 1]
```diff
@@ -483,10 +493,26 @@ where
...
+    // More information is available on the lint's tracking issue
+    // https://github.com/rust-lang/rust/issues/42868.
+    //
+    // Using this `callback` variable is a workaround for the lint's warning.
+    let result;
+    {
+        // Introduce a new scope to limit the duration of the `parser` loan
+        let callback = |input: &mut Parser<'i, '_>| parser.parse_prelude(name, input);
+        result = parse_until_before(input, delimiters, callback);
+    }
```
Comment (SimonSapin) [inline src/rules_and_declarations.rs]: Does the warning happen without assignment outside of `let`, like this?

```rust
let result = {
    …
};
```

If not, please change it and similarly `let prelude;` below.

Comment (lqd) [reply src/rules_and_declarations.rs]: this also works indeed, done

**Pull request merged**